server/status,kvserver: gate child metrics on setting & expose storage metrics for tenants

ajwerner opened the pull request:
See individual commits. 

Touches #47905

Commit:
server/status: gate reporting child metrics behind a cluster setting

It would be bad if all of a sudden these new time series showed up and
started causing existing queries to go wild. Let's add a cluster setting
to gate them.

Release note: None

Commit:
kvserver: track per-tenant storage metrics

This commit separates out the "storage" metrics which correspond to MVCC stats
and adopts the `aggmetric` package. Replicas will report their tenant ID when
they go to update the storage metrics.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51819)
<!-- Reviewable:end -->

ajwerner commented:
cc @imjching

ajwerner commented:
I'm thinking I should split up the second commit into one that adds a tenant ID to the replica and then the second which is about the metrics. I should augment that tenant ID commit to add it to the `RangeInfo` and add some testing there.

Commit:
kvserver: add tenantID field to Replica set on first initialization

This commit adds a field to the Replica which indicates its corresponding
tenant. This field is set from the start key upon first initialization.

Release note: None

ajwerner commented:
@tbg if you're back, you might be interested in this.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> sorry for letting this sit.

> Reviewed 5 of 5 files at r1, 5 of 5 files at r2, 9 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDG3l6s3TEoSZRexdnI:brwmys2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/kv/kvserver/metrics.go, line 1216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG1pc85-9zS3zMARGu:-MDG1pc900SUd0YGgjd6:b-xp8th1) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1216)):*
> ```Go
> 	SysCount        *aggmetric.AggGauge
> 
> 	mu struct {
> ```

Make a note that everything in this struct is invisible to `metric.Struct`.
___
*[pkg/kv/kvserver/metrics.go, line 1218 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG2TPy20WR-uubc9WQ:-MDG2TPy20WR-uubc9WR:b-rxt79h) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1218)):*
> ```Go
> 	mu struct {
> 		syncutil.RWMutex
> 		tenants map[roachpb.TenantID]*tenantStorageMetrics
> ```

Is a `syncutil.IntMap` a good fit here? I would think so, give that the keys should be stable over time. It would be nice to avoid the RLock in ApplyToStateMachine under the Replica mutex.
___
*[pkg/kv/kvserver/metrics.go, line 1227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG1TT1A4gSARHnGXdf:-MDG1TT1A4gSARHnGXdg:bvcp1ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1227)):*
> ```Go
> func (sm *TenantsStorageMetrics) MetricStruct() {}
> 
> func (sm *TenantsStorageMetrics) ensureTenant(tenantID roachpb.TenantID) {
> ```

Give this a comment.

Also, consider changing to `acquireTenant` so that it's clear that it's the counterpart to `releaseTenant`.
___
*[pkg/kv/kvserver/metrics.go, line 1365 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG21ic3YmUDOjn3Kta:-MDG21ic3YmUDOjn3Ktb:b-v5rupx) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1365)):*
> ```Go
> 		LeaseEpochCount:           metric.NewGauge(metaLeaseEpochCount),
> 
> 		// Storage metrics.
> ```

Stale comment.
___
*[pkg/kv/kvserver/replica.go, line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG-m9D-5L__yaBdIw4:-MDG-m9D-5L__yaBdIw5:bc35a6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica.go#L195)):*
> ```Go
> 	log.AmbientContext
> 
> 	// TODO(tschottdorf): Duplicates r.mu.state.desc.RangeID; revisit that.
> ```

You can probably remove this TODO. I don't think we're ever going to address it, or that we necessarily even want to.
___
*[pkg/kv/kvserver/replica.go, line 1059 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG08g5-d2grbjRGTg4:-MDG08g5-d2grbjRGTg5:b-2gkjhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica.go#L1059)):*
> ```Go
> 		// descriptor so we know that it is safe to read here. If the descriptor is
> 		// not initialized, then we'll just leave the tenant ID as zero.
> 		if desc.IsInitialized() {
> ```

You can probably also compare `r.tenantID != (roachpb.TenantID{})` and avoid the other questions here.
___
*[pkg/kv/kvserver/replica_init.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG0Q-n0IxlaLr9dBOV:-MDG0Q-n0IxlaLr9dBOW:b-wc87tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica_init.go#L300)):*
> ```Go
> 
> 	// Initialize the tenant. The must be the first time that the descriptor has
> 	// been initialized.
> ```

Make a note that `desc.StartKey` never changes over the lifetime of a Range.
___
*[pkg/kv/kvserver/replica_init.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG3Of1EJwKmHgX4AZW:-MDG3Of1EJwKmHgX4AZX:b-3bekc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/replica_init.go#L307)):*
> ```Go
> 				"replica %v: %v", r, err)
> 		}
> 		r.store.metrics.ensureTenant(tenantID)
> ```

minor nit: strange ordering here. I'd expect `r.tenantID = tenantID` to be first.
___
*[pkg/kv/kvserver/store.go, line 1486 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG3bwGEYQ-SSDQ4Kmc:-MDG3bwGEYQ-SSDQ4Kmd:b-5cw30d) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/store.go#L1486)):*
> ```Go
> 			// Add this range and its stats to our counter.
> 			s.metrics.ReplicaCount.Inc(1)
> 			s.metrics.addMVCCStats(rep.tenantID, rep.GetMVCCStats())
> ```

`addMVCCStats` will blow up if this tenantID is uninitialized, right?
___
*[pkg/util/metric/prometheus_exporter.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG-9cB7peydOUI1J02:-MDG-9cB7peydOUI1J03:bgnk922) ([raw file](https://github.com/cockroachdb/cockroach/blob/71656a296d1238ca37ec415adf9ad3c7c85b57b5/pkg/util/metric/prometheus_exporter.go#L85)):*
> ```Go
> 			return
> 		}
> 		if promIter, ok := v.(PrometheusIterable); ok {
> ```

nit: maybe flatten this too, since you're already here?

```go
promIter, ok := v.(PrometheusIterable)
if !ok || !includeChildMetrics {
    return
}
...
```


<!-- Sent from Reviewable.io -->

Commit:
aggmetric: add some syntatic sugar around constructing aggmetrics

Often you want to construct a bunch of metrics using the same labels. This
commit adds a `Builder` construct which makes that easy.

Release note: None

ajwerner reviewed (commented):
I added some syntactic sugar to aggmetic and a test around the concurrency stuff for the tenantStorageMetrics. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDIiWGmDYZEn4lrKhTt:b-q4b85d)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/metrics.go, line 1218 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG2TPy20WR-uubc9WQ:-MDHM_dw9ncvaKz1mA5C:brgkm39) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1218)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is a `syncutil.IntMap` a good fit here? I would think so, give that the keys should be stable over time. It would be nice to avoid the RLock in ApplyToStateMachine under the Replica mutex.
</blockquote></details>

Sure, doing this ends up requiring fine-grained synchronization for each tenant. I took an initial stab of doing it without any locking but it got tricky and didn't feel worth it as the only locking is in acquire and release. 
___
*[pkg/kv/kvserver/metrics.go, line 1227 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG1TT1A4gSARHnGXdf:-MDIU6Kd-fkoCtX7fWl_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1227)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this a comment.

Also, consider changing to `acquireTenant` so that it's clear that it's the counterpart to `releaseTenant`.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica.go, line 195 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG-m9D-5L__yaBdIw4:-MDIdzDnCtLkvDln_eJ3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica.go#L195)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You can probably remove this TODO. I don't think we're ever going to address it, or that we necessarily even want to.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica.go, line 1059 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG08g5-d2grbjRGTg4:-MDIeDYV9mW3Q8ty-TaW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica.go#L1059)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You can probably also compare `r.tenantID != (roachpb.TenantID{})` and avoid the other questions here.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_init.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG0Q-n0IxlaLr9dBOV:-MDITwUJ0eBo8uMSBEtl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3432a4b2a62b20b539bdf869219010d7a4700ebc/pkg/kv/kvserver/replica_init.go#L300)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Make a note that `desc.StartKey` never changes over the lifetime of a Range.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_init.go, line 307 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG3Of1EJwKmHgX4AZW:-MDITvq07XaKDaj8PvHi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/replica_init.go#L307)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

minor nit: strange ordering here. I'd expect `r.tenantID = tenantID` to be first.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store.go, line 1486 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG3bwGEYQ-SSDQ4Kmc:-MDITyX6DUhC8UviH7pC:b-387r7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/store.go#L1486)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`addMVCCStats` will blow up if this tenantID is uninitialized, right?
</blockquote></details>

Correct, I now plumbed through a context to get a better fatal error. 
___
*[pkg/util/metric/prometheus_exporter.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG-9cB7peydOUI1J02:-MDIfkRD7Xp7UMeOLLB5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71656a296d1238ca37ec415adf9ad3c7c85b57b5/pkg/util/metric/prometheus_exporter.go#L85)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: maybe flatten this too, since you're already here?

```go
promIter, ok := v.(PrometheusIterable)
if !ok || !includeChildMetrics {
    return
}
...
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 5 files at r1, 14 of 14 files at r4, 5 of 5 files at r5, 2 of 2 files at r6, 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDJs5hC6LlWzZlvExIP:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/kv/kvserver/metrics.go, line 1217 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJqIHR-cZLjkd9ZOZL:-MDJqIHR-cZLjkd9ZOZM:b-oe926z) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1217)):*
> ```Go
> 	SysCount        *aggmetric.AggGauge
> 
> 	// This struct is invisble to the metric package.
> ```

invisble
___
*[pkg/kv/kvserver/metrics.go, line 1224 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJqcyG4ioDEUYciax4:-MDJqcyG4ioDEUYciax5:b6gnaro) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1224)):*
> ```Go
> 
> // MetricStruct makes TenantsStorageMetrics a metric.Struct.
> func (sm *TenantsStorageMetrics) MetricStruct() {}
> ```

`metric.Struct` could have this method unexported, right? Not suggesting you do so, just seems like a minor itch that could be scratched.
___
*[pkg/kv/kvserver/replica.go, line 196 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJp4_B7fCHxDciWo_v:-MDJp4_B7fCHxDciWo_w:b-hue2gh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica.go#L196)):*
> ```Go
> 
> 	RangeID  roachpb.RangeID  // Only set by the constructor
> 	tenantID roachpb.TenantID // Set when first initialized, not modified after
> ```

As a casual reader, this leaves me wondering about the locking. What ensures that the one write to the field isn't racing with reads of it? Surely `.State()` can be called at any given moment.
___
*[pkg/kv/kvserver/replica_init.go, line 308 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJpsEi4MZCPejFwjF4:-MDJpsEi4MZCPejFwjF5:b-laa4fl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica_init.go#L308)):*
> ```Go
> 				"replica %v: %v", r, err)
> 		}
> 		r.tenantID = tenantID
> ```

Make clear on the field that we're intentionally leaving `r.tenantID` unpopulated if it's the system tenant. Or make it populated in this case, which seems to follow the principle of least surprise but might mean jumping through a small hoop to avoid "acquiring tenant metrics" in this case.
___
*[pkg/kv/kvserver/tenantrate/metrics.go, line 68 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJq01cE4DXZx6n38M4:-MDJq01cE4DXZx6n38M5:b7yg96w) ([raw file](https://github.com/cockroachdb/cockroach/blob/85d1c183a05caf35bc0074dedcb0ede8d642e24b/pkg/kv/kvserver/tenantrate/metrics.go#L68)):*
> ```Go
> 
> func makeMetrics() Metrics {
> 	b := aggmetric.MakeBuilder(TenantIDLabel)
> ```

I like the attention to UX here.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
cc @madelineliao - I think you'll catch fallout when you rebase https://github.com/cockroachdb/cockroach/pull/50938 on top of this change (if this one merges earlier). It should be relatively mechanical to resolve it, I hope.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDJsBaGBQOf1K1G0m1V:bwb0s75)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica.go, line 196 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJp4_B7fCHxDciWo_v:-MDKZpPL4QRaTyfo6nNs:be680kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica.go#L196)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

As a casual reader, this leaves me wondering about the locking. What ensures that the one write to the field isn't racing with reads of it? Surely `.State()` can be called at any given moment.
</blockquote></details>

Yeah it’s confusing. I wanted to be able to use it without locking in contexts when the replica is known to be initialized (like during command application). In State() we only access it under the mutex. Initialization happens also with mu held. There was some more commentary in State() before. Let me add some more both on the field and in the method 
___
*[pkg/kv/kvserver/replica_init.go, line 308 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJpsEi4MZCPejFwjF4:-MDKZYfZ7IHSsjZyka07:b-cbicba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica_init.go#L308)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Make clear on the field that we're intentionally leaving `r.tenantID` unpopulated if it's the system tenant. Or make it populated in this case, which seems to follow the principle of least surprise but might mean jumping through a small hoop to avoid "acquiring tenant metrics" in this case.
</blockquote></details>

What do you mean? The above decode call returns the system tenant in cases when it cannot decode a secondary tenant. I’ll add commentary that that’s the case. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica.go, line 196 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJp4_B7fCHxDciWo_v:-MDKmWE460A-aTsl384r:bf63xlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica.go#L196)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Yeah it’s confusing. I wanted to be able to use it without locking in contexts when the replica is known to be initialized (like during command application). In State() we only access it under the mutex. Initialization happens also with mu held. There was some more commentary in State() before. Let me add some more both on the field and in the method 
</blockquote></details>

Are you sure this is worth it? If I read this correctly, the hot path acquires and releases the mutex just before you read this field. Couldn't you just... read it under the mutex? I would love for this to be simpler.
___
*[pkg/kv/kvserver/replica_init.go, line 308 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJpsEi4MZCPejFwjF4:-MDKm-zE2fpxkzr-8MAb:bdkiah) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica_init.go#L308)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What do you mean? The above decode call returns the system tenant in cases when it cannot decode a secondary tenant. I’ll add commentary that that’s the case. 
</blockquote></details>

Oh, my bad. Not even sure that comment is necessary. I just made bogus assumptions about `DecodeTenantPrefix`.


<!-- Sent from Reviewable.io -->

Commit:
kvserver/rangefeed: use RunAsyncTask instead of RunWorker

The lifetime of these calls doesn't actually follow the lifetime of a goroutine
which is going to respond to quiesce. Given that, there's a race which can
occur when a Rangefeed request comes in concurrently with a Stop call.

The test which ensures that the tenant ID is properly set hits this race.

Release note: None

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 14 files at r4, 4 of 5 files at r5, 2 of 2 files at r6, 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDLd_IE6eqZplGu5Izd:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner and @tbg)

___
*[pkg/kv/kvserver/metrics.go, line 1218 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG2TPy20WR-uubc9WQ:-MDLa_2kAOUZj6Z0ZZMW:b-84pzlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1218)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Sure, doing this ends up requiring fine-grained synchronization for each tenant. I took an initial stab of doing it without any locking but it got tricky and didn't feel worth it as the only locking is in acquire and release. 
</blockquote></details>

Thanks. And do you mind adding a comment like `// map[roachpb.TenantID]*tenantStorageMetrics`. We do this in most places that we use IntMaps.
___
*[pkg/kv/kvserver/metrics.go, line 1250 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDLccjN8QMDrUeRmkfV:-MDLccjN8QMDrUeRmkfW:brz3jte) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1250)):*
> ```Go
> 			}
> 			// Somebody else concurrently took the reference count to zero, go back
> 			// around.
> ```

Mention that because of the locking, the map is guaranteed to have been changed by the time we go back around. That alleviates concerns of spinning without making forward progress.
___
*[pkg/kv/kvserver/metrics.go, line 1258 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDLbwsA9CTUGT9kwCWL:-MDLbwsA9CTUGT9kwCWM:b-gor3bf) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1258)):*
> ```Go
> 				// Lost the race with another goroutine to add the instance, go back
> 				// around.
> 				m.mu.Unlock()
> ```

nit: any reason to unlock? It makes it look like this struct is still relevant.
___
*[pkg/kv/kvserver/replica.go, line 196 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJp4_B7fCHxDciWo_v:-MDLd7i27kiwZRam4rrr:bgnle6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica.go#L196)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Are you sure this is worth it? If I read this correctly, the hot path acquires and releases the mutex just before you read this field. Couldn't you just... read it under the mutex? I would love for this to be simpler.
</blockquote></details>

If not that, then maybe put this behind a method and assert under race that `desc.IsInitialized`.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Alright y'all this is about ready to go. I added another mostly unrelated commit to fix the test failures related to races with the stopper and rangefeeds that I tickled in my new test. I think the problem was already there and became more prevalent with the rangefeeds for leases. Usually in testing we shut down all nodes at the same time which is why we don't see it. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDLheODB-0eQVCd_6Rc:b-87aen6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/metrics.go, line 1218 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDG2TPy20WR-uubc9WQ:-MDLh9hb7DrRvUzh771x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32de0adc92068fb68faccfd196fe7aaa3f4a7028/pkg/kv/kvserver/metrics.go#L1218)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Thanks. And do you mind adding a comment like `// map[roachpb.TenantID]*tenantStorageMetrics`. We do this in most places that we use IntMaps.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/metrics.go, line 1217 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJqIHR-cZLjkd9ZOZL:-MDLhDkP7-9PHltV9S4H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1217)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

invisble
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/metrics.go, line 1250 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDLccjN8QMDrUeRmkfV:-MDLhUFo9Bl_2YfqbRCX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1250)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mention that because of the locking, the map is guaranteed to have been changed by the time we go back around. That alleviates concerns of spinning without making forward progress.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/metrics.go, line 1258 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDLbwsA9CTUGT9kwCWL:-MDLhH9u0-3nK9Gr6Mjc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/973e49e254ee9bd83ffba5b3c4472984db2ed719/pkg/kv/kvserver/metrics.go#L1258)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: any reason to unlock? It makes it look like this struct is still relevant.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica.go, line 196 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDJp4_B7fCHxDciWo_v:-MDLhYNh0Zy5LIcspmXj:btfciqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c8022855c0d4b05a7ff99b62ff1eb274a2842b3/pkg/kv/kvserver/replica.go#L196)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If not that, then maybe put this behind a method and assert under race that `desc.IsInitialized`.
</blockquote></details>

I moved it under `mu` 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 16 of 16 files at r8, 1 of 1 files at r9, 5 of 5 files at r10, 2 of 2 files at r11, 10 of 10 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner and @tbg)

___
*[pkg/kv/kvserver/client_replica_test.go, line 3406 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOor4c0z_yMpAYzvvP:-MDOor4c0z_yMpAYzvvQ:b-kg8woe) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b3d462a72fa183c252f81832d590dc5e69b228/pkg/kv/kvserver/client_replica_test.go#L3406)):*
> ```Go
> 	stickySpecTestServerArgs := base.TestServerArgs{
> 		RaftConfig: raftConfig,
> 		Insecure:   true,
> ```

Why? Make it cheaper under race? Add a comment.
___
*[pkg/kv/kvserver/replica_init.go, line 219 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOpGyQ02gzNH8_mApb:-MDOpGyQ02gzNH8_mApc:b-ay9aef) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b3d462a72fa183c252f81832d590dc5e69b228/pkg/kv/kvserver/replica_init.go#L219)):*
> ```Go
> // TenantID returns the associated tenant ID. It will be invalid only if the
> // replica is not initialized.
> func (r *Replica) TenantID() roachpb.TenantID {
> ```

I would feel better about a `(roachpb.TenantID, bool)` signature here. Using the zero tenant value in the wrong places tends to get you a panic, and future callers may not read the comment.
___
*[pkg/kv/kvserver/rangefeed/processor.go, line 172 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOnqUr491WzZlh4E5B:-MDOnqUr491WzZlh4E5C:b226ri5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b69d55e30d0504a358a0db94ab8051ac3b05bf5/pkg/kv/kvserver/rangefeed/processor.go#L172)):*
> ```Go
> func (p *Processor) Start(stopper *stop.Stopper, rtsIter storage.SimpleIterator) {
> 	ctx := p.AnnotateCtx(context.Background())
> 	if err := stopper.RunAsyncTask(ctx, "rangefeed.Processor", func(ctx context.Context) {
> ```

is the race between adding to the WaitGroup and calling .Wait() on it?

One day we're going to clean up the awkward api around workers and tasks.

I think the officlal workaround these days is guarding the call to `RunWorker` in a `RunTask`, I'm not sure I like that though. This is fine.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod @tbg's remaining comments.

> Reviewed 4 of 5 files at r10, 9 of 10 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-:-MDPoqr_1hC9qfh4Mgcl:b-rpdo2m)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @tbg)

___
*[pkg/kv/kvserver/rangefeed/processor.go, line 172 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOnqUr491WzZlh4E5B:-MDPmFK-5u-FBvA2VqU4:b-m3janw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b69d55e30d0504a358a0db94ab8051ac3b05bf5/pkg/kv/kvserver/rangefeed/processor.go#L172)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

is the race between adding to the WaitGroup and calling .Wait() on it?

One day we're going to clean up the awkward api around workers and tasks.

I think the officlal workaround these days is guarding the call to `RunWorker` in a `RunTask`, I'm not sure I like that though. This is fine.
</blockquote></details>

Do you remember what the difference is between a worker and an async task? Just that workers are allowed to run while the stopper is quiesced but not yet stopped? What's the rationale behind that? Did we get this pattern from somewhere?

It seems like we could fix this in a pretty straightforward manner with a change like:
```go
+++ b/pkg/util/stop/stopper.go
@@ -210,8 +210,16 @@ func (s *Stopper) Recover(ctx context.Context) {

 // RunWorker runs the supplied function as a "worker" to be stopped
 // by the stopper. The function <f> is run in a goroutine.
-func (s *Stopper) RunWorker(ctx context.Context, f func(context.Context)) {
+func (s *Stopper) RunWorker(ctx context.Context, f func(context.Context)) error {
+       s.mu.Lock()
+       select {
+       case <-s.stopper:
+               s.mu.Unlock()
+               return ErrUnavailable
+       default:
+       }
        s.stop.Add(1)
+       s.mu.Unlock()
        go func() {
                // Remove any associated span; we need to ensure this because the
                // worker may run longer than the caller which presumably closes
@@ -221,6 +229,7 @@ func (s *Stopper) RunWorker(ctx context.Context, f func(context.Context)) {
                defer s.stop.Done()
                f(ctx)
        }()
+       return nil
 }

 // AddCloser adds an object to close after the stopper has been stopped.
```

But do we want to? There must be a reason to keep the worker vs. task distinction, right?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @tbg)

___
*[pkg/kv/kvserver/client_replica_test.go, line 3406 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOor4c0z_yMpAYzvvP:-MDPkgI4BY7iiaA_2gEL:bbp6saq) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b3d462a72fa183c252f81832d590dc5e69b228/pkg/kv/kvserver/client_replica_test.go#L3406)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Why? Make it cheaper under race? Add a comment.
</blockquote></details>

Yes. Comment added. I was seeing a ton of the following errors when bringing up the second node under stressrace. I was very surprised by them and wonder if they indicate a bug somewhere. There should be 6 seconds for the handshake to complete. Adding insecure made it pass without any problems for a long time. 

```
I200729 04:51:41.425008 3108 kv/kvserver/replica_command.go:1046  [n1,s1,r35/1:/{Tenant/2-Max}] could not promote [n2,s2] to voter, rolling back: snapshot failed: failed to connect to n2 at 127.0.0.1:37291: initial connection heartbeat failed: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: context deadline exceeded"
I200729 04:51:41.626732 3108 kv/kvserver/replica_command.go:1609  [n1,s1,r35/1:/{Tenant/2-Max}] change replicas (add [] remove [(n2,s2):13LEARNER]): existing descriptor r35:/{Tenant/2-Max} [(n1,s1):1, (n2,s2):13LEARNER, next=14, gen=24, sticky=9223372036.854775807,2147483647]
W200729 04:51:42.664122 398 server/node.go:748  [n1,summaries] health alerts detected: {Alerts:[{StoreID:1 Category:METRICS Description:ranges.underreplicated Value:35}]}
W200729 04:51:43.204064 327 kv/kvserver/store_raft.go:502  [n1,s1,r4/1:/System{/tsd-tse}] handle raft ready: 0.8s [applied=1, batches=1, state_assertions=0]
I200729 04:51:43.253910 3108 kv/kvserver/replica_raft.go:247  [n1,s1,r35/1:/{Tenant/2-Max}] proposing SIMPLE(r13) REMOVE_REPLICA[(n2,s2):13LEARNER]: after=[(n1,s1):1] next=14
I200729 04:51:43.655443 3108 kv/kvserver/replica_command.go:1374  [n1,s1,r35/1:/{Tenant/2-Max}] rolled back learner n2,s2 in r35:/{Tenant/2-Max} [(n1,s1):1, (n2,s2):13LEARNER, next=14, gen=24, sticky=9223372036.854775807,2147483647]
I200729 04:51:43.886896 1454 kv/kvserver/client_replica_test.go:3463  SucceedsSoon: AdminChangeReplicas error: snapshot failed: failed to connect to n2 at 127.0.0.1:37291: initial connection heartbeat failed: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: context deadline exceeded"
    TestTenantID/(2)_not_set_before_snapshot: client_replica_test.go:3463: condition failed to evaluate within 45s: AdminChangeReplicas error: snapshot failed: failed to connect to n2 at 127.0.0.1:37291: initial connection heartbeat failed: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: context deadline exceeded"
```
___
*[pkg/kv/kvserver/replica_init.go, line 219 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOpGyQ02gzNH8_mApb:-MDPmw06EJdEnx9-ltOk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01b3d462a72fa183c252f81832d590dc5e69b228/pkg/kv/kvserver/replica_init.go#L219)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I would feel better about a `(roachpb.TenantID, bool)` signature here. Using the zero tenant value in the wrong places tends to get you a panic, and future callers may not read the comment.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/rangefeed/processor.go, line 172 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOnqUr491WzZlh4E5B:-MDPlLXBB9xDwjjjhoqJ:b-df7myo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b69d55e30d0504a358a0db94ab8051ac3b05bf5/pkg/kv/kvserver/rangefeed/processor.go#L172)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

is the race between adding to the WaitGroup and calling .Wait() on it?

One day we're going to clean up the awkward api around workers and tasks.

I think the officlal workaround these days is guarding the call to `RunWorker` in a `RunTask`, I'm not sure I like that though. This is fine.
</blockquote></details>

Exactly, I didn't see how to tie the lifetime of this call cleanly to a `Task`. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r+

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51819)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/rangefeed/processor.go, line 172 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/51819#-MDOnqUr491WzZlh4E5B:-MDQMlRx5tY_p6G7nEtk:b-8qyqbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b69d55e30d0504a358a0db94ab8051ac3b05bf5/pkg/kv/kvserver/rangefeed/processor.go#L172)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Exactly, I didn't see how to tie the lifetime of this call cleanly to a `Task`. 
</blockquote></details>

It's just old and soggy code. I think originally we tried in CRDB to distinguish between "background workers" that keep the system running (like raft message traffic) and short-lived tasks, and we wanted to be able to cleanly drain out tasks before tearing down the "background workers". I think we learned time and time again that this is a fool's errand and that we just need cancellation awareness everywhere. Given more idle time I would revisit whether workers can be removed and replaced by tasks everywhere.


<!-- Sent from Reviewable.io -->
